packages/caver-wallet/src/keyringContainer.js

```diff
@@ -0,0 +1,308 @@
+    signMessage(address, data, role, index) {
+        const keyring = this.getKeyring(address)
+        if (keyring === undefined) throw new Error(`Failed to find keyring from wallet with ${address}`)
+        return keyring.signMessage(data, role, index)
+    }
+
+    /**
+     * signs the transaction using one key and return the transactionHash
+     *
+     * @param {string} address An address of keyring in keyringContainer.
+     * @param {Transaction} transaction A transaction object.
+     * @param {number} [index] An index of key to use for signing.
+     * @param {function} [hasher] A function to return hash of transaction.
+     * @return {string}
+     */
+    async signWithKey(address, transaction, index, hasher) {
+        if (!transaction.from || transaction.from === '0x') transaction.from = address
+        if (transaction.from.toLowerCase() !== address.toLowerCase())
+            throw new Error(`The from address of the transaction is different with the address of the keyring to use.`)
+
+        // Optional parameter processing
+        // (address transaction) / (address transaction index) / (address transaction hasher) / (address transaction index hasher)
+        if (_.isFunction(index) && hasher === undefined) {
+            hasher = index
+            index = 0
+        }
+        if (index === undefined) index = 0
+        if (!_.isNumber(index)) throw new Error(`Invalid index type: ${typeof index}`)
+        if (hasher === undefined) hasher = TransactionHasher.getHashForSigning
+
+        await transaction.fillTransaction()
+        const hash = hasher(transaction)
+        const role = transaction.type.includes('ACCOUNT_UPDATE') ? KEY_ROLE.ROLE_ACCOUNT_UPDATE_KEY : KEY_ROLE.ROLE_TRANSACTION_KEY
+
+        const keyring = this.getKeyring(address)
+        if (keyring === undefined) throw new Error(`Failed to find keyring from wallet with ${address}`)
+        const sig = keyring.signWithKey(hash, transaction.chainId, role, index)
+
+        transaction.appendSignatures(sig)
+
+        return hash
+    }
+
+    /**
+     * signs the transaction using keys and return the transactionHash
+     *
+     * @param {string} address An address of keyring in keyringContainer.
+     * @param {Transaction} transaction A transaction object.
+     * @param {function} [hasher] A function to return hash of transaction.
+     * @return {string}
+     */
+    async signWithKeys(address, transaction, hasher = TransactionHasher.getHashForSigning) {
+        if (!transaction.from || transaction.from === '0x') transaction.from = address
+        if (transaction.from.toLowerCase() !== address.toLowerCase())
+            throw new Error(`The from address of the transaction is different with the address of the keyring to use.`)
+
+        await transaction.fillTransaction()
+        const hash = hasher(transaction)
+        const role = transaction.type.includes('ACCOUNT_UPDATE') ? KEY_ROLE.ROLE_ACCOUNT_UPDATE_KEY : KEY_ROLE.ROLE_TRANSACTION_KEY
```

Comment:
Can we make a function for this? It is used in several places.

Reply:
I made function for this seperately named 'determineRoleToSign'